File: apps/desktop/src/store.ts
/**
 * What survives closing the window.
 *
 * The cache holds answers that were paid for, so re-reading a reference the app
 * has seen before costs nothing. The session holds what was on screen: which
 * references were read and what came back.
 *
 * Neither can be kept by the web view itself, so both live with the host. In a
 * plain browser there is no host, and both fall back to memory — which is what
 * the dev server has always had.
 */

import { invoke } from '@tauri-apps/api/core';
import { ALL_EXTENSIONS, EXTENSIONS, type ResultCache } from '@dialect/core';

const hasHost = (): boolean =>
  typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window;

/**
 * The gateway's cache, kept by the host.
 *
 * A read that fails is a miss rather than an error: a cache that cannot be
 * reached should slow the app down, not stop it.
 */
export class HostCache implements ResultCache {
  readonly #memory = new Map<string, unknown>();

  async get(key: string): Promise<unknown | undefined> {
    if (!hasHost()) return this.#memory.get(key);

    try {
      const text = await invoke<string | null>('cache_get', { key });
      return text === null ? undefined : (JSON.parse(text) as unknown);
    } catch {
      return undefined;
    }
  }

  async set(key: string, value: unknown): Promise<void> {
    if (!hasHost()) {
      this.#memory.set(key, value);
      return;
    }
    try {
      await invoke<void>('cache_set', { key, value: JSON.stringify(value) });
    } catch {
      // Losing a cache write costs money later, not correctness now.
    }
  }
}

/** Whether the window is running inside the app rather than a plain browser. */
export const hasDesktop = (): boolean => hasHost();

export interface CacheStats {
  entries: number;
  bytes: number;
}

export async function cacheStats(): Promise<CacheStats | null> {
  if (!hasHost()) return null;
  try {
    return await invoke<CacheStats>('cache_stats', {});
  } catch {
    return null;
  }
}

export async function clearCache(): Promise<number> {
  memoryLibrary = [];
  if (!hasHost()) return 0;
  return invoke<number>('cache_clear', {});
}

/**
 * The one cache the window uses.
 *
 * A shared instance, not a new one per call: without a host the fallback lives
 * in the instance, so a second HostCache would look into an empty map and
 * report a miss for something already there.
 */
export const hostCache = new HostCache();

/** Read an answer straight out of the cache, for a reference already paid for. */
export async function cachedAnswer(key: string): Promise<unknown | undefined> {
  return hostCache.get(key);
}

// ---------------------------------------------------------------------------
// The session
// ---------------------------------------------------------------------------

/**
 * Bumped to 2 when the panels went: a session used to hold the open document,
 * the chosen target and the batch list, and now holds none of them. Where the
 * work is lives in the graph, which is a file of its own.
 */
export const SESSION_VERSION = 2 as const;

/**
 * What survives a restart that belongs to nothing else.
 *
 * One thing: what has been spent. A cap that resets when the window closes is
 * not a cap, and this is the only reason the file still exists.
 */
export interface Session {
  version: typeof SESSION_VERSION;
  spentUsd: number;
}

let memorySession: Session | undefined;

export async function loadSession(): Promise<Session | undefined> {
  if (!hasHost()) return memorySession;

  try {
    const text = await invoke<string | null>('session_get', {});
    if (text === null) return undefined;

    const parsed = JSON.parse(text) as Session;
    // A session written by an older build is discarded rather than half-read.
    return parsed.version === SESSION_VERSION ? parsed : undefined;
  } catch {
    return undefined;
  }
}

export async function saveSession(session: Session): Promise<void> {
  if (!hasHost()) {
    memorySession = session;
    return;
  }
  try {
    await invoke<void>('session_set', { value: JSON.stringify(session) });
  } catch {
    // Not worth interrupting anyone over; the work itself is unaffected.
  }
}

export async function clearSession(): Promise<void> {
  memorySession = undefined;
  if (hasHost()) await invoke<void>('session_clear', {}).catch(() => undefined);
}

// ---------------------------------------------------------------------------
// The library: what the cache holds, in terms a person recognises
// ---------------------------------------------------------------------------

/**
 * A cache entry is keyed by a hash and holds only an answer — no name, no
 * picture, nothing anyone could pick out of a list. This is the index over
 * them, so a reference read last week can be found instead of paid for twice.
 */
export interface LibraryEntry {
  /** The cache key the answer is filed under. */
  key: string;
  /** What the file was called. */
  ref: string;
  /** ISO timestamp of the read that paid for it. */
  readAt: string;
  /** A small JPEG data URL, so it can be recognised on sight. */
  thumb?: string;
}

let memoryLibrary: LibraryEntry[] = [];

export async function loadLibrary(): Promise<LibraryEntry[]> {
  if (!hasHost()) return memoryLibrary;
  try {
    const parsed: unknown = JSON.parse(await invoke<string>('library_get', {}));
    return Array.isArray(parsed) ? (parsed as LibraryEntry[]) : [];
  } catch {
    return [];
  }
}

/** Newest first, one entry per key: re-reading a reference updates its date. */
export async function rememberRead(entry: LibraryEntry): Promise<LibraryEntry[]> {
  const next = [entry, ...(await loadLibrary()).filter((e) => e.key !== entry.key)];

  if (!hasHost()) memoryLibrary = next;
  else {
    try {
      await invoke<void>('library_set', { value: JSON.stringify(next) });
    } catch {
      // The answer is still cached; only its label is lost.
    }
  }
  return next;
}

/** A small picture is worth keeping; a large one is not. */
export async function thumbnail(file: File, max = 240): Promise<string | undefined> {
  try {
    const bitmap = await createImageBitmap(file);
    const scale = Math.min(1, max / Math.max(bitmap.width, bitmap.height));
    const width = Math.max(1, Math.round(bitmap.width * scale));
    const height = Math.max(1, Math.round(bitmap.height * scale));

    const canvas = document.createElement('canvas');
    canvas.width = width;
    canvas.height = height;
    canvas.getContext('2d')?.drawImage(bitmap, 0, 0, width, height);
    bitmap.close();

    return canvas.toDataURL('image/jpeg', 0.7);
  } catch {
    return undefined;
  }
}

// ---------------------------------------------------------------------------
// Saving prompts
// ---------------------------------------------------------------------------

export interface OutFile {
  name: string;
  contents: string;
}

/**
 * Ask where, write there, and offer to show it.
 *
 * Returns the folder, or null if the dialog was dismissed — the caller says
 * nothing in that case, because cancelling is not a failure.
 */
export async function savePromptsTo(files: OutFile[]): Promise<string | null> {
  if (!hasHost()) {
    throw new Error('Saving needs the desktop app: a browser cannot choose a folder to write to.');
  }

  const { open } = await import('@tauri-apps/plugin-dialog');
  const dir = await open({ directory: true, title: 'Where should the prompts go?' });
  if (typeof dir !== 'string') return null;

  await invoke<number>('save_prompts', { dir, files });
  return dir;
}

// showFolder moved to folders.ts, where the permission it needs is asked for.
export { showFolder } from './folders.ts';

// ---------------------------------------------------------------------------
// Reading a video
// ---------------------------------------------------------------------------

export interface MediaTools {
  ffmpeg: boolean;
  ffprobe: boolean;
}

export interface VideoProbe {
  duration_s: number;
  width: number;
  height: number;
  fps: number;
  has_audio: boolean;
  aspect_ratio: string;
}

export interface VideoFrame {
  at: number;
  base64: string;
}

/** Whether a video can be read here at all, so the window can say so up front. */
export async function mediaTools(): Promise<MediaTools> {
  if (!hasHost()) return { ffmpeg: false, ffprobe: false };
  try {
    return await invoke<MediaTools>('media_tools', {});
  } catch {
    return { ffmpeg: false, ffprobe: false };
  }
}

/**
 * References come in by path, not by drop.
 *
 * A dropped file reaches the page as bytes with no name on disk, and handing a
 * few hundred megabytes across as base64 just to have ffmpeg read it would be
 * absurd. So files are chosen, and only paths travel — for every kind, because
 * one button cannot take three kinds if one of them arrives differently.
 */
export async function pickReferences(imagesOnly = false): Promise<string[]> {
  if (!hasHost()) {
    throw new Error('Choosing files needs the desktop app.');
  }

  const { open } = await import('@tauri-apps/plugin-dialog');
  const picked = await open({
    multiple: true,
    title: 'Choose references',
    // Without ffmpeg a video cannot be read at all, so it is not offered rather
    // than accepted and refused later.
    filters: imagesOnly
      ? [{ name: 'Images', extensions: [...EXTENSIONS.image] }]
      : [
          { name: 'References', extensions: [...ALL_EXTENSIONS] },
          { name: 'Images', extensions: [...EXTENSIONS.image] },
          { name: 'Video', extensions: [...EXTENSIONS.video] },
          { name: 'Audio', extensions: [...EXTENSIONS.audio] },
        ],
  });

  if (typeof picked === 'string') return [picked];
  return Array.isArray(picked) ? picked.filter((p): p is string => typeof p === 'string') : [];
}

/** A folder of references, listed by the host so every kind arrives by path. */
export async function pickFolder(): Promise<string | null> {
  if (!hasHost()) {
    throw new Error('Choosing a folder needs the desktop app.');
  }

  const { open } = await import('@tauri-apps/plugin-dialog');
  const dir = await open({ directory: true, title: 'Choose a folder of references' });
  return typeof dir === 'string' ? dir : null;
}

export interface Found {
  name: string;
  path: string;
  bytes: number;
}

export const scanFolder = (dir: string): Promise<Found[]> =>
  invoke<Found[]>('folder_scan', { dir });

/** Bytes off disk, for an image the host picked rather than the page received. */
export const readFile = (path: string): Promise<{ base64: string; bytes: number }> =>
  invoke<{ base64: string; bytes: number }>('file_read', { path });

/**
 * Something to recognise a reference by, whatever kind it is: the image, the
 * video's first frame, or the audio's spectrogram.
 */
export async function thumbOf(path: string): Promise<string | undefined> {
  if (!hasHost()) return undefined;
  try {
    return `data:image/jpeg;base64,${await invoke<string>('file_thumb', { path })}`;
  } catch {
    // A reference with no picture is still a reference.
    return undefined;
  }
}

export const probeVideo = (path: string): Promise<VideoProbe> =>
  invoke<VideoProbe>('media_probe', { path });

export const videoFrames = (path: string, count = 5): Promise<VideoFrame[]> =>
  invoke<VideoFrame[]>('media_frames', { path, count });

// ---------------------------------------------------------------------------
// Reading audio
// ---------------------------------------------------------------------------

export interface AudioTags {
  duration_s: number;
  sample_rate: number;
  channels: number;
  title: string | null;
  artist: string | null;
  genre: string | null;
}

export interface Measured {
  probe: AudioTags;
  tempo: { bpm: number; confidence: number } | null;
  key: { name: string; confidence: number } | null;
  lufs: number | null;
  lra: number | null;
  /** A spectrogram and a waveform, in that order. */
  pictures: Array<{ kind: string; base64: string }>;
}

export const measureAudio = (path: string): Promise<Measured> =>
  invoke<Measured>('audio_measure', { path });

File: apps/desktop/src-tauri/src/store.rs
//! What survives closing the window.
//!
//! Two things, for two different reasons.
//!
//! The cache holds answers that were paid for. Keyed by the hash of the image
//! and the question, so re-reading a reference the app has seen before costs
//! nothing — which is the whole reason it exists, and the reason it has to
//! outlive the process.
//!
//! The session holds what the window was showing: which references were read
//! and what came back. The files themselves cannot be kept — a dropped file is
//! gone once the page reloads — so this is a record, not a resumable job.
//!
//! The library is what makes the cache usable. Cache entries are keyed by a
//! hash and hold only an answer: no name, no picture, nothing a person could
//! recognise. The library is the index over them — what each one was called,
//! when it was read, and a thumbnail small enough to keep — so a reference read
//! last week can be found and brought back instead of paid for twice.

use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

/// The gateway's keys are sha256 hex. Anything else is not ours to write, and
/// is the shape a path traversal would arrive in.
fn checked_key(key: &str) -> Result<&str, String> {
    let ok = key.len() == 64 && key.bytes().all(|b| b.is_ascii_hexdigit());
    if ok {
        Ok(key)
    } else {
        Err(format!("Not a cache key: {key}"))
    }
}

fn data_dir(app: &AppHandle) -> Result<PathBuf, String> {
    app.path()
        .app_data_dir()
        .map_err(|e| format!("No place to keep data: {e}"))
}

fn cache_dir(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = data_dir(app)?.join("cache");
    fs::create_dir_all(&dir).map_err(|e| format!("Could not make the cache directory: {e}"))?;
    Ok(dir)
}

/// Everything below takes a directory rather than an `AppHandle`, so it can be
/// tested against a real one. The commands only resolve where that is.
pub fn read_entry(dir: &Path, key: &str) -> Result<Option<String>, String> {
    let path = dir.join(format!("{}.json", checked_key(key)?));
    match fs::read_to_string(&path) {
        Ok(text) => Ok(Some(text)),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(format!("Could not read a cached answer: {e}")),
    }
}

pub fn write_entry(dir: &Path, key: &str, value: &str) -> Result<(), String> {
    let path = dir.join(format!("{}.json", checked_key(key)?));

    // Written aside and renamed, so a crash mid-write leaves the old entry or
    // none — never half of one, which would fail to parse forever.
    let temp = dir.join(format!("{}.{}.tmp", checked_key(key)?, std::process::id()));
    fs::write(&temp, value).map_err(|e| format!("Could not write a cached answer: {e}"))?;
    fs::rename(&temp, &path).map_err(|e| format!("Could not store a cached answer: {e}"))
}

#[tauri::command]
pub fn cache_get(app: AppHandle, key: String) -> Result<Option<String>, String> {
    read_entry(&cache_dir(&app)?, &key)
}

#[tauri::command]
pub fn cache_set(app: AppHandle, key: String, value: String) -> Result<(), String> {
    write_entry(&cache_dir(&app)?, &key, &value)
}

#[derive(Serialize)]
pub struct CacheStats {
    entries: u64,
    bytes: u64,
}

/// What the cache is holding, so the settings panel can say what clearing costs.
pub fn stats_in(dir: &Path) -> Result<CacheStats, String> {
    let mut stats = CacheStats { entries: 0, bytes: 0 };

    for entry in fs::read_dir(dir).map_err(|e| format!("Could not read the cache: {e}"))? {
        let entry = entry.map_err(|e| format!("Could not read the cache: {e}"))?;
        if entry.path().extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        if let Ok(meta) = entry.metadata() {
            stats.entries += 1;
            stats.bytes += meta.len();
        }
    }
    Ok(stats)
}

#[tauri::command]
pub fn cache_stats(app: AppHandle) -> Result<CacheStats, String> {
    stats_in(&cache_dir(&app)?)
}

pub fn clear_in(dir: &Path) -> Result<u64, String> {
    let mut removed = 0;

    for entry in fs::read_dir(dir).map_err(|e| format!("Could not read the cache: {e}"))? {
        let entry = entry.map_err(|e| format!("Could not read the cache: {e}"))?;
        if entry.path().extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        if fs::remove_file(entry.path()).is_ok() {
            removed += 1;
        }
    }
    Ok(removed)
}

#[tauri::command]
pub fn cache_clear(app: AppHandle) -> Result<u64, String> {
    let removed = clear_in(&cache_dir(&app)?)?;
    // The index without the answers it points at is worse than neither.
    write_library(&library_path(&app)?, "[]")?;
    Ok(removed)
}

// ---------------------------------------------------------------------------
// The library: what the cache is holding, in terms a person recognises
// ---------------------------------------------------------------------------

fn library_path(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = data_dir(app)?;
    fs::create_dir_all(&dir).map_err(|e| format!("Could not make the data directory: {e}"))?;
    Ok(dir.join("library.json"))
}

pub fn read_library(path: &Path) -> Result<String, String> {
    match fs::read_to_string(path) {
        Ok(text) => Ok(text),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok("[]".to_string()),
        Err(e) => Err(format!("Could not read the library: {e}")),
    }
}

pub fn write_library(path: &Path, value: &str) -> Result<(), String> {
    let temp = path.with_extension("tmp");
    fs::write(&temp, value).map_err(|e| format!("Could not write the library: {e}"))?;
    fs::rename(&temp, path).map_err(|e| format!("Could not store the library: {e}"))
}

#[tauri::command]
pub fn library_get(app: AppHandle) -> Result<String, String> {
    read_library(&library_path(&app)?)
}

#[tauri::command]
pub fn library_set(app: AppHandle, value: String) -> Result<(), String> {
    write_library(&library_path(&app)?, &value)
}

// ---------------------------------------------------------------------------
// The session
// ---------------------------------------------------------------------------

fn session_path(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = data_dir(app)?;
    fs::create_dir_all(&dir).map_err(|e| format!("Could not make the data directory: {e}"))?;
    Ok(dir.join("session.json"))
}

pub fn read_session(path: &Path) -> Result<Option<String>, String> {
    match fs::read_to_string(path) {
        Ok(text) => Ok(Some(text)),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(format!("Could not read the last session: {e}")),
    }
}

pub fn write_session(path: &Path, value: &str) -> Result<(), String> {
    let temp = path.with_extension("tmp");
    fs::write(&temp, value).map_err(|e| format!("Could not write the session: {e}"))?;
    fs::rename(&temp, path).map_err(|e| format!("Could not store the session: {e}"))
}

pub fn remove_session(path: &Path) -> Result<(), String> {
    match fs::remove_file(path) {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(format!("Could not clear the session: {e}")),
    }
}

#[tauri::command]
pub fn session_get(app: AppHandle) -> Result<Option<String>, String> {
    read_session(&session_path(&app)?)
}

#[tauri::command]
pub fn session_set(app: AppHandle, value: String) -> Result<(), String> {
    write_session(&session_path(&app)?, &value)
}

#[tauri::command]
pub fn session_clear(app: AppHandle) -> Result<(), String> {
    remove_session(&session_path(&app)?)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn scratch(name: &str) -> PathBuf {
        let dir = std::env::temp_dir().join(format!("dialect-store-{}-{name}", std::process::id()));
        let _ = fs::remove_dir_all(&dir);
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    const KEY: &str = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
    const OTHER: &str = "fedcba9876543210fedcba9876543210fedcba9876543210fedcba9876543210";

    #[test]
    fn only_accepts_the_shape_the_gateway_produces() {
        assert!(checked_key(KEY).is_ok());

        // The bad shapes include the one a path traversal would arrive in.
        for bad in ["", "abc", "../../etc/passwd", &"g".repeat(64), &"a".repeat(63)] {
            assert!(checked_key(bad).is_err(), "{bad} should have been refused");
        }
    }

    #[test]
    fn an_answer_survives_being_written_and_read_back() {
        let dir = scratch("roundtrip");

        assert_eq!(read_entry(&dir, KEY).unwrap(), None);

        write_entry(&dir, KEY, r#"{"headline":"a bottle"}"#).unwrap();
        assert_eq!(
            read_entry(&dir, KEY).unwrap().as_deref(),
            Some(r#"{"headline":"a bottle"}"#)
        );

        // Overwriting keeps the latest, not both.
        write_entry(&dir, KEY, r#"{"headline":"a different bottle"}"#).unwrap();
        assert!(read_entry(&dir, KEY).unwrap().unwrap().contains("different"));

        fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn leaves_no_temporary_file_behind() {
        let dir = scratch("temp");
        write_entry(&dir, KEY, "{}").unwrap();

        let names: Vec<String> = fs::read_dir(&dir)
            .unwrap()
            .map(|e| e.unwrap().file_name().to_string_lossy().into_owned())
            .collect();

        assert_eq!(names, vec![format!("{KEY}.json")]);
        fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn a_session_survives_being_written_and_read_back() {
        let dir = scratch("session");
        let path = dir.join("session.json");

        // Nothing yet is not an error; it is a first run.
        assert_eq!(read_session(&path).unwrap(), None);

        write_session(&path, r#"{"version":1,"target":"nano-banana-2"}"#).unwrap();
        assert!(read_session(&path).unwrap().unwrap().contains("nano-banana-2"));

        // The next write replaces it rather than appending.
        write_session(&path, r#"{"version":1,"target":"kling-3-omni"}"#).unwrap();
        let text = read_session(&path).unwrap().unwrap();
        assert!(text.contains("kling-3-omni"));
        assert!(!text.contains("nano-banana-2"));

        remove_session(&path).unwrap();
        assert_eq!(read_session(&path).unwrap(), None);
        // Clearing something already gone is not an error.
        remove_session(&path).unwrap();

        fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn only_writes_plain_file_names() {
        for good in ["a.txt", "ref-001_nano-banana-2.txt", "shot 3.ir.json"] {
            assert!(checked_name(good).is_ok(), "{good} should have been allowed");
        }
        // The web view builds these from a naming pattern, which is exactly
        // where a stray separator or a parent reference would come from.
        for bad in ["", "..", "../out.txt", "a/b.txt", "a\\b.txt", "C:evil.txt"] {
            assert!(checked_name(bad).is_err(), "{bad} should have been refused");
        }
    }

    #[test]
    fn writes_a_whole_batch_into_a_folder_it_makes() {
        let dir = scratch("save").join("prompts");

        let files = vec![
            OutFile { name: "one.txt".into(), contents: "first prompt".into() },
            OutFile { name: "one.ir.json".into(), contents: "{}".into() },
        ];
        assert_eq!(write_all(&dir, &files).unwrap(), 2);

        assert_eq!(fs::read_to_string(dir.join("one.txt")).unwrap(), "first prompt");
        assert!(dir.join("one.ir.json").exists());

        // One bad name refuses the write rather than half-writing the batch.
        let bad = vec![OutFile { name: "../escape.txt".into(), contents: "x".into() }];
        assert!(write_all(&dir, &bad).is_err());

        fs::remove_dir_all(dir.parent().unwrap()).unwrap();
    }

    #[test]
    fn the_library_starts_empty_rather_than_missing() {
        let dir = scratch("library");
        let path = dir.join("library.json");

        // A first run has no file, and that is an empty library, not an error.
        assert_eq!(read_library(&path).unwrap(), "[]");

        write_library(&path, r#"[{"key":"abc","ref":"one.png"}]"#).unwrap();
        assert!(read_library(&path).unwrap().contains("one.png"));

        write_library(&path, "[]").unwrap();
        assert_eq!(read_library(&path).unwrap(), "[]");

        fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn counts_and_clears_what_it_holds() {
        let dir = scratch("stats");
        write_entry(&dir, KEY, "{\"a\":1}").unwrap();
        write_entry(&dir, OTHER, "{\"b\":2}").unwrap();

        let stats = stats_in(&dir).unwrap();
        assert_eq!(stats.entries, 2);
        assert!(stats.bytes > 0);

        assert_eq!(clear_in(&dir).unwrap(), 2);
        assert_eq!(stats_in(&dir).unwrap().entries, 0);
        assert_eq!(read_entry(&dir, KEY).unwrap(), None);

        fs::remove_dir_all(&dir).unwrap();
    }
}

// ---------------------------------------------------------------------------
// Saving prompts
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct OutFile {
    name: String,
    contents: String,
}

/// Refuses anything that is not a plain file name. The web view chooses these
/// from a naming pattern, and a pattern is exactly where a stray `..` or a
/// second directory separator would come from.
fn checked_name(name: &str) -> Result<&str, String> {
    let bad = name.is_empty()
        || name.contains("..")
        || name.contains('/')
        || name.contains('\\')
        || name.contains(':');
    if bad {
        Err(format!("Not a file name: {name}"))
    } else {
        Ok(name)
    }
}

pub fn write_all(dir: &Path, files: &[OutFile]) -> Result<u64, String> {
    fs::create_dir_all(dir).map_err(|e| format!("Could not use that folder: {e}"))?;

    let mut written = 0;
    for file in files {
        let path = dir.join(checked_name(&file.name)?);
        fs::write(&path, &file.contents)
            .map_err(|e| format!("Could not write {}: {e}", file.name))?;
        written += 1;
    }
    Ok(written)
}

#[tauri::command]
pub fn save_prompts(dir: String, files: Vec<OutFile>) -> Result<u64, String> {
    write_all(Path::new(&dir), &files)
}
